File: PHP/src/Parrot.php
<?php

namespace Parrot;

abstract class Parrot {
  /** @var int ParrotTypeEnum */
  private $type;
  /** @var int */
  private $numberOfCoconuts = 0;
  /** @var  double */
  private $voltage;
  /** @var  boolean */
  private $isNailed;

  /**
   * Parrot constructor.
   * @param int $type
   * @param int $numberOfCoconuts
   * @param float $voltage
   * @param bool $isNailed
   */
  public function __construct($type, $numberOfCoconuts, $voltage, $isNailed) {
    $this->type = $type;
    $this->numberOfCoconuts = $numberOfCoconuts;
    $this->voltage = $voltage;
    $this->isNailed = $isNailed;
  }

  /**
   * @return float
   * @throws \Exception
   */
  abstract public function getSpeed();

  protected function getBaseSpeed() {
    return 12.0;
  }

}

File: Swift-2.2/src/Parrot.swift
//
//  Parrot.swift
//  ParrotSwift
//

import Foundation

public class Parrot {
	
	private let type: ParrotTypeEnum
	private let numberOfCoconuts: Int
	private let voltage: Double
	private let isNailed: Bool
	
	
	public init(_type: ParrotTypeEnum, numberOfCoconuts: Int = 0, voltage: Double, isNailed: Bool) {
		self.type = _type
		self.numberOfCoconuts = numberOfCoconuts
		self.voltage = voltage
		self.isNailed = isNailed
	}
	
	public func getSpeed() -> Double {
		switch(type) {
		case .EUROPEAN:
			return getBaseSpeed()
		case .AFRICAN:
			return max(0, getBaseSpeed() - getLoadFactor() * Double(numberOfCoconuts))
		case .NORWEGIAN_BLUE:
			return (isNailed) ? 0 : getBaseSpeed(voltage)
		}
	
		fatalError("Should be unreachable")
	}
	
	private func getBaseSpeed(voltage: Double) -> Double {
		return min(24.0, voltage*getBaseSpeed())
	}
	
	private func getLoadFactor() -> Double {
		return 9.0
	}
	
	private func getBaseSpeed() -> Double {
		return 12.0
	}
	
	
}
